cache/pmdk_cache.cc

```diff
@@ -0,0 +1,622 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include "cache/pmdk_cache.h"
+
+#include <cassert>
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+
+#include "util/mutexlock.h"
+
+// TODO: make these run-time variables
+// #define PHEAP_PATH "/mnt/pmem/pmdk_cache/"
+#define PHEAP_PATH "/dev/shm/pmdk_cache/"
+#define PMEMOBJ_POOL_SIZE ((size_t)024 * 1024 * 1024 * 2)
+
+namespace ROCKSDB_NAMESPACE {
+
+PMDKCacheShard::PMDKCacheShard(size_t capacity, bool strict_capacity_limit,
+                               double /*high_pri_pool_ratio*/,
+                               bool use_adaptive_mutex,
+                               CacheMetadataChargePolicy metadata_charge_policy,
+                               size_t persist_capacity, size_t shard_id)
+    : capacity_(capacity),
+      persistent_capacity_(persist_capacity),
+      strict_capacity_limit_(strict_capacity_limit),
+      mutex_(use_adaptive_mutex) {
+  set_metadata_charge_policy(metadata_charge_policy);
+
+  // TODO: set up an LRUCacheShard as transient tier.
+
+  // Set up persistent memory pool (pop)
+  std::string heap_file = PHEAP_PATH + std::to_string(shard_id);
+  // TODO: use an alternative to access() that works on all platforms.
+  // TODO: provide option to ignore existing pool and create new one.
+  // TDOO: persist static metadata (capacity, etc) and report inconsistency with
+  // arguments during recovery.
+  try {
+    if (access(heap_file.c_str(), F_OK) != 0) {
+      pop_ = po::pool<PersistentRoot>::create(heap_file, "pmdk_cache_pool",
+                                              PMEMOBJ_POOL_SIZE, S_IRWXU);
+      po::transaction::run(pop_, [&] {
+        auto root = pop_.root();
+        root->persistent_hashtable =
+            po::make_persistent<PersistTierHashTable>(&pop_);
+        root->persistent_lru_list = po::make_persistent<PersistentEntry>();
+        root->usage = 0;
+        root->lru_usage = 0;
+        root->era = 0;
+        usage_ = po::persistent_ptr<size_t>(&root->usage.get_rw());
+        lru_usage_ = po::persistent_ptr<size_t>(&root->lru_usage.get_rw());
+        persistent_hashtable_ = root->persistent_hashtable;
+        lru_ = root->persistent_lru_list;
+        // empty circular linked lru list
+        lru_->next_lru = lru_;
+        lru_->prev_lru = lru_;
+      });
+      era_ = 0;
+    } else {
+      pop_ = po::pool<PersistentRoot>::open(heap_file, "pmdk_cache_pool");
+      persistent_hashtable_ = pop_.root()->persistent_hashtable;
+      persistent_hashtable_->SetPop(&pop_);
+      usage_ = po::persistent_ptr<size_t>(&pop_.root()->usage.get_rw());
+      lru_usage_ = po::persistent_ptr<size_t>(&pop_.root()->lru_usage.get_rw());
+      lru_ = pop_.root()->persistent_lru_list;
+      era_ = ++pop_.root()->era;
+    }
+  } catch (...) {
+    // TODO: quit here.
```

Comment:
We should not terminate the process because cache initialization fails. There are multiple options.
1. Make constructor exception-free, and use another `Initialize()` to do the heavy-lifting that currently exists in `PMDKCacheShard::PMDKCacheShard()`. The `Initialize()` function handles exceptions and returns a status code.
2. Throw exception from `PMDKCacheShard::PMDKCacheShard()`, handle this exception in `NewPMDKCache()` function. The exception can be turned into a status code that will be returned to the caller. Also make sure other parts of RocksDB cannot call `PMDKCacheShard` directly.

Reply:
`NewPMDKCache()` returns a pointer to the newly-initialized cache. Will returning a nullptr work?